[WIP] Refac generator and add alov generator

sallamander opened the pull request:
@dansbecker What I've got so far and what I'm working on.

sallamander committed:
Stash code for commenting

sallamander committed:
Move second parsing of imagenet bb

Previously, the parsing of the imagnet bb down to only those bounding
boxes corresponding to succesfully downloaded images was built into the
image generator. This meant that it was getting run each time we were
working through the generator, when in reality we only need to run it
once (after we have the final set of images). This parsing (e.g. the
second parsing of the imagenet bounding boxes) is now in its own script.

sallamander committed:
Setup structure for a master image generator

sallamander committed:
Adjust makefile for moving bad images

dansbecker commented on Makefile:
```diff
@@ -62,19 +62,27 @@ work/imagenet/parsed_bb.csv: work/.imagenet_training_bbox_sentinel \
 	python motion_tracker/data_setup/parse_imagenet_bb.py data/imagenet/bb/ \
 		work/imagenet/parsed_bb.csv
 
+# Run after images are downloaded, to filter out those images 
```
Normally with make, I don't expect to keep track of when to run things. Anything we need this for will depend on it.  

And anything that has to be run before this can be run automatically when we try to build this.  

Is this a special case?

sallamander replied:
Are you just referencing the comment, or what's the issue with this? If it's the comment, happy to remove that. 

We can't have this run until imagenet images are downloaded and we know what failed, so it's dependent on that. But,I don't think that makes it a special case.

dansbecker replied:
yeah. I'd just remove the comment.  At first, it made me wonder if there's something more than meets the eye (e.g. a dependency we couldn't represent in the Makefile).

sallamander replied:
Done.

sallamander committed:
Fill in and debug imagenet generator in master

sallamander committed:
Add viz capability for generator

sallamander commented on motion_tracker/data_setup/parse_imagenet_bb2.py:
```diff
@@ -0,0 +1,44 @@
+    raw_image_dir = sys.argv[1]
+    parsed_bb_path = sys.argv[2]
+
+    parse_imagenet_bb(raw_image_dir, parsed_bb_path)
```
Might not need to be it's own file, but for now I thought it made sense. Definitely open to trying to put it somewhere else, but I think it's okay here, too. It might make it a pain to try to play around with that `66%` parameter, though.

dansbecker replied:
We should rename this from bb2.  That makes it unclear whether it's half of all the bb's, and should be combined with bb1... or whether it is the next step (which it actually is).  Perhaps call it `usable_bbs.csv` or `filtered_bbs.csv`

Let's save that for a separate PR though rather than doing it here.

sallamander commented on motion_tracker/utils/crop_vis.py:
```diff
@@ -1,7 +1,7 @@
 import cv2
-from motion_tracker.utils.image_cropping import Coords, crop_and_resize
+from image_cropping import Coords, crop_and_resize
 
```
So, watch out for this if you try to run this stuff. For some reason this wasn't working for me even when running from the main directory of the repo. Not sure if I don't have something setup that you do, or what. But, I had to revert to that import for it to work.

dansbecker replied:
We need to figure something out on this.  It was working for me, and after this change it no longer runs for me (unless I run things from the utils directory.)

Can you add the repository's root directory to your pythonpath and see if the previous import works?

sallamander commented on motion_tracker/utils/crop_vis.py:
```diff
@@ -41,3 +41,18 @@ def show_stages_of_random_crop(img, box_coords, output_width=256, output_height=
+                box_coords[2], box_coords[3])
+    img_w_bb = add_box(img, bb_coords)
+    show_img(title, img_w_bb)
+
```
This is used at the bottom of the `image_generator.py` to give a kind of slideshow of random croppings for a given image.

sallamander commented on motion_tracker/utils/image_generator.py:
```diff
@@ -0,0 +1,126 @@
+    mas_gen = master_generator() 
+    imgs, boxes = next(mas_gen)
+    for img, box in zip(imgs[0], boxes[0]): 
+        show_single_stage(img, box)
```
You should be able to test out the imagenet generator with this. Right now the alov one is commented out because it's not built. But, I'm somewhat pumped that this appears to be working. It generates 11 images per image right now - the original + 10 random croppings.

sallamander committed:
Update parse_alov_bb.py for easy use in generator

Before this commit, parse_alov_bb.py was built to simply parse all of the bounding box files for the alov videos and store the result in a csv. It is much more helpful and efficient in the image generator that will be used in the net to have the parsing actually set up pairs of the video frames that will be fed into the net. This commit sets that up.

sallamander commented on motion_tracker/data_setup/parse_alov_bb.py:
```diff
@@ -28,45 +28,86 @@ def parse_file(bbox_dir, filename):
+    there is nothing after it. 
+    """
+    pass
+
```
This function and the one right above are where I'm planning on doing the merging/joining so that the end result of the `.csv` is that each row holds information relating to the two frames that are fed into the net together.

dansbecker commented on motion_tracker/data_setup/parse_imagenet_bb2.py:
```diff
@@ -0,0 +1,44 @@
+"""A short script for futher filtering the alov bounding boxes.
```
I think this script is misnamed (having the name imagenet in it)

sallamander replied:
Yep, sorry that's just a bad comment. Will fix.

dansbecker commented on motion_tracker/utils/crop_vis.py:
```diff
@@ -41,3 +41,18 @@ def show_stages_of_random_crop(img, box_coords, output_width=256, output_height=
     add_box(final_img, box_coords)
     show_img("Random Crop", final_img)
+
+def show_single_stage(img, box_coords, title="Happy Dance Image"): 
```
After the current PR is merged, I'd like to get rid of the current show_img, and replace it with something like this.  We can give box_coords a default argument of None, and only plot the box if box_coords isn't None.

That's basically this function... and having it obviates the need for add_box and show_img.

It also simplifies a bunch of places where I sequentially called add_box and show_img.

It's not like we are ever calling add_box and then not showing the image :)

sallamander replied:
Cool. I thought the same thing, and had started adjusted the function directly above that but wanted to hold off.

dansbecker commented:
The output of the generator needs to have the same format as the previous imagenet_generator had.  That is, a dictionary where each key is a string describing it's contents, and each value is a numpy array of the relevant data, with the first dimension being the observation level.

dansbecker commented on motion_tracker/utils/image_generator.py:
```diff
@@ -0,0 +1,151 @@
+        video_pairs = 0
+        images = 0
+        num_obs = 0
+        # Might need to add in some check or something in the 
```
While it's possible one of the groups of crops will get cut off, that doesn't seem like a big deal, and I think the current code implements this well enough.

dansbecker commented on motion_tracker/utils/image_generator.py:
```diff
@@ -0,0 +1,151 @@
+        boxes_lst = []
+        img_row = img_metadata.sample(1)
+        
+        # Make sure to read in the image only once, and calculate 
```
One of the costs of long comments is that they increase vertical spacing in the code, making you scroll around more when trying to read multiple parts of the file. 

It sounds like a small cost, but navigating between two functions can be distracting when you are trying to think about the two of them together.

You can reduce this cost by making the comment lines longer (continuing further to the right)

sallamander replied:
Sorry, yeah I'll make sure to go out further. For now, those comments were somewhat just to highlight what I meant to do there, and I was planning on removing some of them after everything was said and done.

dansbecker replied:
Awesome.

On Sat, Apr 30, 2016 at 7:58 AM, Sean Sall notifications@github.com wrote:

> In motion_tracker/utils/image_generator.py
> https://github.com/dansbecker/motion-tracking/pull/14#discussion_r61669965
> :
> 
> > -    -------
> > -        imgs_lst: list of images
> > -        boxes_lst: list of bounding box coords
> > -    """
> >   +
> > -    raw_image_dir = 'data/imagenet/images/'
> > -    img_metadata = pd.read_csv('work/imagenet/parsed_bb2.csv')
> > -    output_width = 256
> > -    output_height = 256
> >   +
> > -    while True:
> > -        imgs_lst = []
> > -        boxes_lst = []
> > -        img_row = img_metadata.sample(1)
> >   +
> > -        # Make sure to read in the image only once, and calculate
> 
> Sorry, yeah I'll make sure to go out further. For now, those comments were
> somewhat just to highlight what I meant to do there, and I was planning on
> removing some of them after everything was said and done.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/dansbecker/motion-tracking/pull/14/files/ae85f5ab63ad24107fab1fa228800e0fcd08cd06#r61669965

dansbecker committed:
Change output format of generators for consistency with fit_generator

dansbecker committed:
Mixor bug fix

dansbecker committed:
Fix bookkeeping bugs in image generators

sallamander committed:
Merge pull request #15 from dansbecker/db_mods_to_refac_generator

Change output format of generators for consistency with fit_generator

sallamander committed:
Finish up parse_alov_bb.py

sallamander committed:
Update image_generator.py with alov generator

sallamander committed:
Refactor and cleanup alov generator code

sallamander commented on motion_tracker/utils/image_generator.py:
```diff
@@ -0,0 +1,193 @@
+                     'start_box': empty_box_accumulator,
+                     'end_img': empty_img_accumulator,
+                     'end_box': empty_box_accumulator}
+        while my_output['start_img'].shape[0] < batch_size:
```
I think this is the only remaining issue here. We need the imagenet generator and the alov generator to generate 10 image pairs each, rather than the batch size. If we generate the batch size, then I think the way that the master generator is built, we'll end up training on only alov or imagenet for each batch, and not both. I understand we want to keep this in a similar format as the master generator, so I was hesitant to switch the `< batch_size` part here. It'd be easy if we wanted to to have some bool passed in that notes whether the generator is being used alone (and this line should use batch size) or as part of the master generator (and this line should be using crops_per_image).

dansbecker replied:
Ah, this part should be easy.  Just change master_generator so it asks for batches of 10 from the imagenet and alov generators (lines 25 and 26).  

It alternates between the generators until it gets up to its desired batch size.

And if we ever want more images from one of the sources than the other, we can specify that there too.

sallamander commented on motion_tracker/utils/image_generator.py:
```diff
@@ -0,0 +1,193 @@
+        for field in my_output:
+            my_output[field] = my_output[field][:batch_size]
+        yield my_output
+
```
A class refactor will definitely be nice here.

sallamander commented on motion_tracker/utils/image_generator.py:
```diff
@@ -0,0 +1,193 @@
+    batch = next(my_gen)
+    for i in range(batch['start_img'].shape[0]):
+        show_single_stage(batch['start_img'][i], batch['start_box'][i])
+        show_single_stage(batch['end_img'][i], batch['end_box'][i])
```
So.... this should work. I was able to see 10 sets of an image pair (e.g. 20 images) before seeing another 10 sets of an image pair, and they corresponded to both `alov` and `imagenet` (but I did have to switch the `< batch_size`  in the individual generators to `< crops_per_image`). 😆 

I didn't make it to any actual refactoring of the image_cropping file, so that's still on my list. After working through it some in trying to debug today, I don't know that there's a lot to do. Also, given the errors we've run into, I think it could be worthwhile in the future/long term to throw some asserts/checks in there that give warnings as to what's wrong. For example, at one point I was getting cropping coords that were greater than the image, and so the cropping coords were restricted to be the max of the image -1. This is definitely the right thing to do coding wise, but it might be nice to throw some kind of warning in there to note that that is going on. It also might be something that is much easier to find if you've worked with images before, leaving any kind of warnings less useful.
